src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+[Spec: Abstract Operations](https://tc39.es/ecma262/#sec-abstract-operations)
+
+## Internal slots and internal methods
+
+**Internal slots** and **internal methods** use names enclosed in \[\[ \]\].
+
+Internal slots are (assumed) data members (of a JavaScript object or a spec-internal data type) and internal methods are (assumed) member functions (of a JavaScript object or a spec-internal data type).
+
+Internal slots and methods are used in the algorithms described by the spec. Internal slots are used for storing the state of the object, and internal methods are functions associated with the object.
+
+[Spec: Object internal methods and internal slots](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots)
+
+For example, every JavaScript object has an internal slot \[\[Prototype\]\] and an internal method \[\[GetOwnProperty\]\].
+
+Internal slots and methods are not accessible from JavaScript, for example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.
+
+Each internal method delegates to a similarly-named abstract operation:
+
+> \[\[GetOwnProperty\]\](P)
+>
+> When the \[\[GetOwnProperty\]\] internal method of O is called with property key P, the following steps are taken:
+>
+> Return ! OrdinaryGetOwnProperty(O, P).
+
+[Spec: \[\[GetOwnProperty\]\]](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-getownproperty-p)
+
+(We'll find out what the exclamation mark means in the next chapter.)
+
+_OrdinaryGetOwnProperty_ is not an internal method, since it's not associated with any object; instead, the object it operates on is passed as a parameter.
+
+_OrdinaryGetOwnProperty_ is called "ordinary" since it operates on ordinary objects. ECMAScript objects can be either **ordinary** or **exotic**. Ordinary objects must have the default behavior for a set of methods called **essential internal methods**. If an object deviates from the default behavior, it's exotic.
+
+The most well-known exotic object is the _Array_, since its length property behaves in a non-default way: setting the length property can remove elements from the _Array_.
+
+Essential internal methods are the methods listed [here](https://tc39.es/ecma262/#table-5).
+
+## Completion Records
+
+What about the question marks and exclamation marks? To understand them, we need to look into **Completion Records**!
+
+A Completion Record is a "record" (or a "struct", if you're more familiar with that term) defined by the spec. A record is a data type which has a fixed set of named fields. C-like languages have structs - JavaScript doesn't, since you can always add properties into an object.
```

Comment:
I think the "struct" comparison is more confusing than not. The spec certainly doesn't care about the memory layout of a Record's fields, for instance (or even has a notion of memory layout). I think it would be fine to simply define what ecma262 considers Records to be in this paragraph.